File: zig/allergies/allergies.zig
const std = @import("std");
const enums = std.enums;
const EnumSet = std.EnumSet;

pub const Allergen = enum(u8) {
    eggs = 1 << 0,
    peanuts = 1 << 1,
    shellfish = 1 << 2,
    strawberries = 1 << 3,
    tomatoes = 1 << 4,
    chocolate = 1 << 5,
    pollen = 1 << 6,
    cats = 1 << 7,
};

pub fn isAllergicTo(score: u8, allergen: Allergen) bool {
    return score & @intFromEnum(allergen) > 0;
}

pub fn initAllergenSet(score: usize) EnumSet(Allergen) {
    var ans = EnumSet(Allergen).initEmpty();
    for (enums.values(Allergen)) |allergen| {
        if (isAllergicTo(@truncate(score), allergen)) ans.insert(allergen);
    }
    return ans;
}

File: cpp/allergies/allergies.cpp
#include "allergies.h"

namespace allergies {
bool allergies::is_allergic_to(string allergen) const {
    return (this->items.count(allergen) != 0);
}

unordered_set<string> allergies::get_allergies() const {
    return (this->items);
}

allergies::allergies(int score) {
    this->items.clear();
    for (size_t i = 0; i < ALLERGENS.size(); i++) {
        if (score & (1 << i)) {
            this->items.insert(ALLERGENS[i]);
        }
    }
}

allergies allergy_test(int score) {
    return allergies(score);
}
}  // namespace allergies
